File: src/xuezh/core/retention.py
from __future__ import annotations

import os
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from pathlib import Path

from xuezh.core import paths

RETENTION_DEFAULTS = {
    "artifacts": 90,
    "backups": 30,
    "exports": 180,
    "cache": 180,
}

RETENTION_ENV = {
    "artifacts": "XUEZH_RETENTION_ARTIFACTS_DAYS",
    "backups": "XUEZH_RETENTION_BACKUPS_DAYS",
    "exports": "XUEZH_RETENTION_EXPORTS_DAYS",
    "cache": "XUEZH_RETENTION_CACHE_DAYS",
}


@dataclass(frozen=True)
class RetentionConfig:
    artifacts_days: int
    backups_days: int
    exports_days: int
    cache_days: int


def load_retention_config() -> RetentionConfig:
    def _days(key: str) -> int:
        env_key = RETENTION_ENV[key]
        raw = os.environ.get(env_key)
        if raw is None:
            return RETENTION_DEFAULTS[key]
        return int(raw)

    return RetentionConfig(
        artifacts_days=_days("artifacts"),
        backups_days=_days("backups"),
        exports_days=_days("exports"),
        cache_days=_days("cache"),
    )


def _files_under(root: Path, subdir: str) -> list[Path]:
    base = root / subdir
    if not base.exists():
        return []
    return sorted([p for p in base.rglob("*") if p.is_file()], key=lambda p: str(p))


def _is_older_than(path: Path, cutoff: datetime) -> bool:
    mtime = datetime.fromtimestamp(path.stat().st_mtime, tz=timezone.utc)
    return mtime < cutoff


def collect_gc_candidates(root: Path, *, now: datetime | None = None) -> list[Path]:
    now = now or datetime.now(timezone.utc)
    config = load_retention_config()

    candidates: list[Path] = []
    dir_windows = {
        "artifacts": config.artifacts_days,
        "backups": config.backups_days,
        "exports": config.exports_days,
        "cache": config.cache_days,
    }

    for subdir, days in dir_windows.items():
        cutoff = now - timedelta(days=days)
        for path in _files_under(root, subdir):
            if _is_older_than(path, cutoff):
                # Resolve relative to workspace and ensure safety.
                candidates.append(paths.resolve_in_workspace(path))

    # Deterministic ordering by workspace-relative path.
    return sorted(candidates, key=lambda p: str(p))

File: internal/xuezh/retention/retention.go
package retention

import (
	"os"
	"path/filepath"
	"sort"
	"time"

	"github.com/joshp123/xuezh/internal/xuezh/paths"
)

type Config struct {
	ArtifactsDays int
	BackupsDays   int
	ExportsDays   int
	CacheDays     int
}

var defaults = map[string]int{
	"artifacts": 90,
	"backups":   30,
	"exports":   180,
	"cache":     180,
}

func LoadConfig() Config {
	return Config{
		ArtifactsDays: defaults["artifacts"],
		BackupsDays:   defaults["backups"],
		ExportsDays:   defaults["exports"],
		CacheDays:     defaults["cache"],
	}
}

func CollectGCCandidates(root string, now time.Time) ([]string, error) {
	config := LoadConfig()
	windows := map[string]int{
		"artifacts": config.ArtifactsDays,
		"backups":   config.BackupsDays,
		"exports":   config.ExportsDays,
		"cache":     config.CacheDays,
	}

	candidates := []string{}
	for subdir, days := range windows {
		cutoff := now.Add(-time.Duration(days) * 24 * time.Hour)
		files, err := filesUnder(root, subdir)
		if err != nil {
			return nil, err
		}
		for _, path := range files {
			info, err := os.Stat(path)
			if err != nil {
				continue
			}
			if info.ModTime().UTC().Before(cutoff) {
				resolved, err := paths.ResolveInWorkspace(path)
				if err != nil {
					continue
				}
				candidates = append(candidates, resolved)
			}
		}
	}
	return sortByRelative(root, candidates), nil
}

func filesUnder(root, subdir string) ([]string, error) {
	base := filepath.Join(root, subdir)
	info, err := os.Stat(base)
	if err != nil {
		if os.IsNotExist(err) {
			return []string{}, nil
		}
		return nil, err
	}
	if !info.IsDir() {
		return []string{}, nil
	}
	pathsOut := []string{}
	err = filepath.WalkDir(base, func(path string, entry os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if entry.IsDir() {
			return nil
		}
		pathsOut = append(pathsOut, path)
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(pathsOut)
	return pathsOut, nil
}

func sortByRelative(root string, candidates []string) []string {
	rootAbs, err := filepath.Abs(root)
	if err != nil {
		return candidates
	}
	withRel := make([]struct {
		abs string
		rel string
	}, 0, len(candidates))
	for _, path := range candidates {
		rel, err := filepath.Rel(rootAbs, path)
		if err != nil {
			rel = path
		}
		withRel = append(withRel, struct {
			abs string
			rel string
		}{abs: path, rel: rel})
	}
	sort.Slice(withRel, func(i, j int) bool {
		return withRel[i].rel < withRel[j].rel
	})
	out := make([]string, 0, len(withRel))
	for _, item := range withRel {
		out = append(out, item.abs)
	}
	return out
}
